File: apiserver/internal/cron/cron.go
// Package cron implements the minimal 5-field cron next-fire computation used
// by /api/schedules. It is a faithful port of server/app/cron.py so the Go
// schedule surface and the Python schedule worker agree on the next fire time.
//
// Field semantics (matching cron.py):
//   - * / lists (a,b) / ranges (a-b) / steps (*/n, a-b/n)
//   - Day-of-week uses 0=Sunday..6=Saturday
//   - When BOTH day-of-month and day-of-week are restricted, BOTH must match
//     (some cron dialects OR them; for a * in one field behaviour is identical)
package cron

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

const searchCap = 1_000_000

type schedule struct {
	minutes map[int]bool
	hours   map[int]bool
	days    map[int]bool
	months  map[int]bool
	dows    map[int]bool
}

func parseField(text string, low, high int) (map[int]bool, error) {
	values := map[int]bool{}
	for _, part := range strings.Split(text, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			return nil, fmt.Errorf("empty cron field part in %q", text)
		}
		step := 1
		if i := strings.Index(part, "/"); i >= 0 {
			s, err := strconv.Atoi(part[i+1:])
			if err != nil || s <= 0 {
				return nil, fmt.Errorf("invalid step in cron field %q", text)
			}
			step = s
			part = part[:i]
		}
		var baseLow, baseHigh int
		switch {
		case part == "" || part == "*":
			baseLow, baseHigh = low, high
		case strings.Contains(part, "-"):
			left, right, ok := strings.Cut(part, "-")
			l, err1 := strconv.Atoi(left)
			r, err2 := strconv.Atoi(right)
			if !ok || err1 != nil || err2 != nil {
				return nil, fmt.Errorf("invalid cron field %q", text)
			}
			baseLow, baseHigh = l, r
		default:
			v, err := strconv.Atoi(part)
			if err != nil {
				return nil, fmt.Errorf("invalid cron field %q", text)
			}
			baseLow, baseHigh = v, v
		}
		for v := baseLow; v <= baseHigh; v += step {
			if low <= v && v <= high {
				values[v] = true
			}
		}
	}
	if len(values) == 0 {
		return nil, fmt.Errorf("invalid cron field %q", text)
	}
	return values, nil
}

// Parse parses a 5-field cron expression.
func Parse(expression string) (*schedule, error) {
	fields := strings.Fields(expression)
	if len(fields) != 5 {
		return nil, fmt.Errorf("cron expression must have 5 fields, got %d: %q", len(fields), expression)
	}
	s := &schedule{}
	var err error
	if s.minutes, err = parseField(fields[0], 0, 59); err != nil {
		return nil, err
	}
	if s.hours, err = parseField(fields[1], 0, 23); err != nil {
		return nil, err
	}
	if s.days, err = parseField(fields[2], 1, 31); err != nil {
		return nil, err
	}
	if s.months, err = parseField(fields[3], 1, 12); err != nil {
		return nil, err
	}
	// Cron DOW 0=Sunday..6=Saturday equals Go's time.Weekday directly, so no
	// conversion (Python's cron.py converts to Python weekday 0=Monday).
	if s.dows, err = parseField(fields[4], 0, 6); err != nil {
		return nil, err
	}
	return s, nil
}

// minValue returns the smallest value in the set (the set is never empty).
func minValue(values map[int]bool) int {
	m := -1
	for v := range values {
		if m == -1 || v < m {
			m = v
		}
	}
	return m
}

// nextInSet returns (smallest value >= current, wrapped) like cron.py.
func nextInSet(values map[int]bool, current int) (int, bool) {
	wrap := minValue(values)
	best := -1
	for v := range values {
		if v >= current && (best == -1 || v < best) {
			best = v
		}
	}
	if best != -1 {
		return best, false
	}
	return wrap, true
}

func jumpDay(t time.Time) time.Time {
	return time.Date(t.Year(), t.Month(), t.Day()+1, 0, 0, 0, 0, t.Location())
}

func jumpMonth(t time.Time) time.Time {
	return time.Date(t.Year(), t.Month()+1, 1, 0, 0, 0, 0, t.Location())
}

// NextAfter returns the first matching datetime strictly after moment, in the
// same location. Mirrors CronSchedule.next_after in cron.py.
func (s *schedule) NextAfter(moment time.Time) (time.Time, error) {
	candidate := time.Date(
		moment.Year(), moment.Month(), moment.Day(), moment.Hour(), moment.Minute(), 0, 0,
		moment.Location(),
	).Add(time.Minute)
	for range searchCap {
		if !s.months[int(candidate.Month())] {
			candidate = jumpMonth(candidate)
			continue
		}
		if !(s.days[candidate.Day()] && s.dows[int(candidate.Weekday())]) {
			candidate = jumpDay(candidate)
			continue
		}
		if !s.hours[candidate.Hour()] {
			hour, wrapped := nextInSet(s.hours, candidate.Hour())
			minute := minValue(s.minutes)
			candidate = time.Date(candidate.Year(), candidate.Month(), candidate.Day(), hour, minute, 0, 0, candidate.Location())
			if wrapped {
				candidate = jumpDay(candidate)
				candidate = time.Date(candidate.Year(), candidate.Month(), candidate.Day(), hour, minute, 0, 0, candidate.Location())
			}
			continue
		}
		if !s.minutes[candidate.Minute()] {
			minute, wrapped := nextInSet(s.minutes, candidate.Minute())
			candidate = time.Date(candidate.Year(), candidate.Month(), candidate.Day(), candidate.Hour(), minute, 0, 0, candidate.Location())
			if wrapped {
				candidate = candidate.Add(time.Hour)
				candidate = time.Date(candidate.Year(), candidate.Month(), candidate.Day(), candidate.Hour(), minute, 0, 0, candidate.Location())
			}
			continue
		}
		return candidate, nil
	}
	return time.Time{}, fmt.Errorf("no matching cron fire within search cap")
}

// NextScheduleFire computes the next cron fire time in the schedule's timezone
// and normalizes it to UTC for storage. Mirrors next_schedule_fire in cron.py.
func NextScheduleFire(expression, timezoneName string, after time.Time) (time.Time, error) {
	loc, err := time.LoadLocation(timezoneName)
	if err != nil {
		return time.Time{}, err
	}
	s, err := Parse(expression)
	if err != nil {
		return time.Time{}, err
	}
	nextLocal, err := s.NextAfter(after.In(loc))
	if err != nil {
		return time.Time{}, err
	}
	return nextLocal.UTC(), nil
}

File: server/app/cron.py
"""Minimal 5-field cron parser and next-fire computation.

Supports the standard fields ``minute hour day-of-month month day-of-week``
with ``*``, lists (``a,b``), ranges (``a-b``) and steps (``*/n`` / ``a-b/n``).
Day-of-week uses 0=Sunday..6=Saturday. Kept dependency-free so the schedule
worker and the web API never need ``croniter``.

Semantics simplification: when BOTH day-of-month and day-of-week are
restricted, this parser requires BOTH to match (some cron dialects OR them).
For the common ``*``-in-one-field cases the behaviour is identical.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone


def _parse_field(text: str, low: int, high: int) -> set[int]:
    values: set[int] = set()
    for part in text.split(","):
        part = part.strip()
        if not part:
            raise ValueError(f"empty cron field part in {text!r}")
        step = 1
        if "/" in part:
            part, step_text = part.split("/", 1)
            step = int(step_text)
            if step <= 0:
                raise ValueError(f"invalid step in cron field {text!r}")
        if part in {"", "*"}:
            base_low, base_high = low, high
        elif "-" in part:
            left, right = part.split("-", 1)
            base_low, base_high = int(left), int(right)
        else:
            base_low = base_high = int(part)
        for value in range(base_low, base_high + 1, step):
            if low <= value <= high:
                values.add(value)
    if not values:
        raise ValueError(f"invalid cron field {text!r}")
    return values


def _next_in_set(values: set[int], current: int) -> tuple[int, bool]:
    """Return (smallest value >= current, wrapped) from a non-empty set."""
    ordered = sorted(values)
    for value in ordered:
        if value >= current:
            return value, False
    return ordered[0], True


class CronSchedule:
    def __init__(self, expression: str):
        fields = expression.split()
        if len(fields) != 5:
            raise ValueError(
                f"cron expression must have 5 fields, got {len(fields)}: {expression!r}"
            )
        self.minutes = _parse_field(fields[0], 0, 59)
        self.hours = _parse_field(fields[1], 0, 23)
        self.days = _parse_field(fields[2], 1, 31)
        self.months = _parse_field(fields[3], 1, 12)
        # Cron day-of-week uses 0=Sunday..6=Saturday; Python weekday() uses
        # 0=Monday..6=Sunday. Convert to Python's convention at parse time.
        self.dows = {
            (value - 1) % 7 for value in _parse_field(fields[4], 0, 6)
        }

    def next_after(self, moment: datetime) -> datetime:
        """Return the first matching datetime strictly after ``moment``."""
        candidate = moment.replace(second=0, microsecond=0) + timedelta(minutes=1)
        for _ in range(1_000_000):
            if candidate.month not in self.months:
                candidate = _jump_month(candidate)
                continue
            if not (candidate.day in self.days and candidate.weekday() in self.dows):
                candidate = _jump_day(candidate)
                continue
            if candidate.hour not in self.hours:
                hour, wrapped = _next_in_set(self.hours, candidate.hour)
                candidate = candidate.replace(hour=hour, minute=min(self.minutes))
                if wrapped:
                    candidate = _jump_day(candidate).replace(
                        hour=hour, minute=min(self.minutes)
                    )
                continue
            if candidate.minute not in self.minutes:
                minute, wrapped = _next_in_set(self.minutes, candidate.minute)
                candidate = candidate.replace(minute=minute)
                if wrapped:
                    candidate += timedelta(hours=1)
                    candidate = candidate.replace(minute=minute)
                continue
            return candidate
        raise ValueError("no matching cron fire within search cap")


def _jump_day(moment: datetime) -> datetime:
    return (moment + timedelta(days=1)).replace(hour=0, minute=0)


def _jump_month(moment: datetime) -> datetime:
    if moment.month == 12:
        return datetime(moment.year + 1, 1, 1, 0, 0)
    return datetime(moment.year, moment.month + 1, 1, 0, 0)


def next_after(expression: str, moment: datetime) -> datetime:
    return CronSchedule(expression).next_after(moment)


def next_schedule_fire(
    expression: str,
    timezone_name: str,
    after: datetime,
) -> datetime:
    """Compute the next cron fire time in the schedule's timezone.

    The cron expression is interpreted in ``timezone_name`` and the returned
    datetime is normalized to UTC for storage.
    """
    from zoneinfo import ZoneInfo

    tz = ZoneInfo(timezone_name)
    local_after = after.astimezone(tz)
    next_local = CronSchedule(expression).next_after(local_after)
    return next_local.astimezone(timezone.utc)
